Stream external references

nicksnyder opened the pull request:
Based on https://github.com/sourcegraph/language-server-protocol/blob/master/extension-workspace-references.md

Also related to https://github.com/sourcegraph/sourcegraph/issues/4340

felixfbecker commented on src/vs/editor/contrib/referenceSearch/common/referenceSearch.ts:
```diff
@@ -39,4 +39,30 @@ export function provideReferences(model: IReadOnlyModel, position: Position, pro
+			if (Array.isArray(result)) {
+				return <IReferenceInformation[]>result;
+			}
+		}, err => {
```
Do you intend to only catch errors of `provideExternalReferences()`, not errors inside the `then` handler? If not, use `.catch()`

nicksnyder replied:
same, just copied code

felixfbecker commented on src/vs/editor/contrib/referenceSearch/common/referenceSearch.ts:
```diff
@@ -39,4 +39,30 @@ export function provideReferences(model: IReadOnlyModel, position: Position, pro
+				return <IReferenceInformation[]>result;
+			}
+		}, err => {
+			onUnexpectedExternalError(err);
```
you can pass this function directly without wrapping it in an arrow function

nicksnyder replied:
same, just copied code

felixfbecker commented on src/vs/editor/contrib/referenceSearch/common/referenceSearch.ts:
```diff
@@ -39,4 +39,30 @@ export function provideReferences(model: IReadOnlyModel, position: Position, pro
+
+	// collect references from all providers
+	const promises = ExternalReferenceProviderRegistry.ordered(model).map(provider => {
+		return asWinJsPromise((token) => {
```
omit parens

felixfbecker commented on src/vs/editor/contrib/referenceSearch/common/referenceSearch.ts:
```diff
@@ -39,4 +39,30 @@ export function provideReferences(model: IReadOnlyModel, position: Position, pro
+		return asWinJsPromise((token) => {
+			return provider.provideExternalReferences(query, hints, token, progress);
+		}).then(result => {
+			if (Array.isArray(result)) {
```
Why this check? If it is not an array, return `undefined`?

nicksnyder replied:
This function is copied from `provideReferences` above, which is written by the vs code team. There may be a subtle reason for this, or this may be completely pointless code, I don't really know.

In the interest of getting something working, my strategy is to just copy exactly what vscode has done.

nicksnyder replied:
Ah, I think this code exists because `provider.provideExternalReferences(query, hints, token, progress)` gets serialized to the extension host and the response is serialized back. We expect certain types here, but ultimately the main thread receives JSON from a potentially untrusted extension.

felixfbecker replied:
I would always adapt the typing to reflect this so tsc actually forces you to do these checks. But I guess forcing a good coding style on the VS Code codebase is out of scope

felixfbecker commented on src/vs/editor/contrib/referenceSearch/common/referenceSearch.ts:
```diff
@@ -39,4 +39,30 @@ export function provideReferences(model: IReadOnlyModel, position: Position, pro
+	return TPromise.join(promises).then(references => {
+		let result: IReferenceInformation[] = [];
+		for (let ref of references) {
+			if (ref) {
```
The array is typed as `ReferenceInformation[][]`, so why this check?

nicksnyder replied:
This does seem redundant given the array check above.

nicksnyder replied:
nevermind, the then above may return undefined if there is not an array. going to leave this as it.

felixfbecker commented on src/vs/editor/contrib/referenceSearch/common/referenceSearch.ts:
```diff
@@ -39,4 +39,30 @@ export function provideReferences(model: IReadOnlyModel, position: Position, pro
+
+	return TPromise.join(promises).then(references => {
+		let result: IReferenceInformation[] = [];
+		for (let ref of references) {
```
2x `const`

felixfbecker commented on src/vs/editor/contrib/referenceSearch/common/referenceSearch.ts:
```diff
@@ -39,4 +39,30 @@ export function provideReferences(model: IReadOnlyModel, position: Position, pro
+		let result: IReferenceInformation[] = [];
+		for (let ref of references) {
+			if (ref) {
+				result.push(...ref);
```
is `ref` long?
If yes, do a second loop so the items don't get pushed on the call stack for perf reasons and argument limit of 65535

nicksnyder replied:
65k would be a ridiculous number of references to send back, but I think it is a good idea to be defensive.

felixfbecker commented on src/vs/vscode.d.ts:
```diff
@@ -1949,6 +1949,48 @@ declare module 'vscode' {
+	 *
+	 * In contrast to `SymbolInformation`, `SymbolDescriptor` includes more concrete,
+	 * language-specific, metadata about the symbol.
+	 */
```
this has a different docblock than https://github.com/sourcegraph/vscode/pull/5/files#diff-2ad890a42cb36d54a33f89dfea7330f6R372?

nicksnyder replied:
monaco.d.ts is generated. I tried copying over the comments but they just get overwritten on next generation, but not sure why they don't get picked up.

felixfbecker replied:
Yeah, comments should be persisted by tsc. Which is why I wondered why the comments are different - maybe you forgot to compile in the end?

nicksnyder replied:
Oh yea, I got confused. I can fix this

felixfbecker commented on src/vs/vscode.d.ts:
```diff
@@ -1949,6 +1949,48 @@ declare module 'vscode' {
+	export interface ExternalReferenceProvider {
+
+		/**
+		 * Impements https://github.com/sourcegraph/language-server-protocol/blob/master/extension-workspace-references.md
```
typo

felixfbecker commented on src/vs/workbench/api/node/extHostLanguageFeatures.ts:
```diff
@@ -214,6 +214,33 @@ class ReferenceAdapter {
+
+	provideExternalReferences(query: modes.ISymbolDescriptor, hints: { [hint: string]: any }, progress: (references: modes.IReferenceInformation[]) => void): TPromise<modes.IReferenceInformation[]> {
+		const convertReferences = (references: vscode.ReferenceInformation[]) => {
+			if (Array.isArray(references)) {
```
`references` is typed as an array, so why the check?

nicksnyder replied:
Same reason as above I think. Sanity check for things received over IPC.

felixfbecker commented on src/vs/workbench/api/node/extHostLanguageFeatures.ts:
```diff
@@ -214,6 +214,33 @@ class ReferenceAdapter {
+	}
+
+	provideExternalReferences(query: modes.ISymbolDescriptor, hints: { [hint: string]: any }, progress: (references: modes.IReferenceInformation[]) => void): TPromise<modes.IReferenceInformation[]> {
+		const convertReferences = (references: vscode.ReferenceInformation[]) => {
```
type the return type as `ReferenceInformation[] | undefined`

nicksnyder replied:
kind of pointless. the type system "knows" `references` is an array so it concludes that this method can never return undefined. Even with the explicit `| undefined` the type of convertReferences is always just `IReferenceInformation[]`

felixfbecker replied:
Yeah, you're right. But imo if `references` can be something else than an array, it should be typed that way... Unexpectedly returning `undefined` is what causes applications to crash with null pointers. But again, I am probably just trying to enforce a coding style that you just can't achieve in the VS Code codebase

nicksnyder replied:
> But imo if references can be something else than an array, it should be typed that way...

This is how the vscode team has chosen to represent their IPC communication. Each end appears strongly typed, but also does a few sanity checks because ultimately it is just serialized json. It isn't really worth our time trying to debate their decisions unless you want to debate them on the upstream repo and submit a pull request.

The most important rule about code style is to be consistent with the code around you, unless you are willing to change all the code around you (whether that means in the function/file/repo). Since we are trying to minimize diffs to reduce conflicts, I think being consistent is the right approach unless there is an actual problem.

Being consistent has the added benefit of if/when we rebase, it will be easier to tell what has changed so we can update our custom functions to match. For example, if `provideReferences` changes, I will update `provideExternalReferences` to match.

felixfbecker replied:
Yes, totally agree with consistency over style :)

Regarding rebasing: This isn't a big difference and unrelated to the discussion, but I wouldn't rebase on upstream VS Code's master. That would rewrite our history, we would lose the ability to track when a change was made. For a long-living branch like this let's _merge_ VS Code's upstream with proper merge commits so the commit history is maintained. This is how we handle this in the LSP fork too

nicksnyder replied:
I was using the term "rebase" loosely, but I am actually not convinced either way as far as rebasing vs merge. It is definitely a discussion worth having when we get to that point. I can see arguments for both sides.

> we would lose the ability to track when a change was made.

What do you mean? I have already done an interactive rebase on vscode/patch once to squash some commits together (to reduce cumulative diff) and it appears to have maintained the original author and commit date.

<img width="508" alt="screen shot 2017-02-26 at 3 40 50 pm" src="https://cloud.githubusercontent.com/assets/754768/23344933/fe0ddfb4-fc39-11e6-8e6a-b1b33dfb8b98.png">

felixfbecker replied:
> I was using the term "rebase" loosely, but I am actually not convinced either way as far as rebasing vs merge. It is definitely a discussion worth having when we get to that point. I can see arguments for both sides.

Yeah, let's discuss this in-depth when we get there. I just wanted to throw this in before we just do it either way :)

> What do you mean? I have already done an interactive rebase on vscode/patch once to squash some commits together (to reduce cumulative diff) and it appears to have maintained the original author and commit date.

git differentiates between _commiter_ vs _author_ and _commit date_ vs _author date_. So when squashing you won't notice (especially on GitHub which shows everything linear). Squashing is great for removing useless commits like "Fix linting error", "Fix typo", etc. I love squash-merging PRs.
But when you do a bunch of commits over a period of multiple weeks (as results of multiple squash-merged PRs), and at the same time the VS Code team makes a lot of commits on upstream, if you then rebase all your history on VS Code's commit history, you loose the information that these developments/commit histories happened in parallel until they were literally "merged" at a specific point. Maybe not when looking at the timestamps, but from looking at the parent relationship between commits. Eg `git log --graph` would not be able to show you two parallel histories anymore that got merged at one or more points, it just shows you a linear history with our commits on the top, rewritten every time we need upstream changes.
Besides that, a rebase requires a force push and every developer needs to do a hard reset instead of a pull, loosing local commits if not done carefully. This is fine and I do it a lot for "own" feature branches, but on a branch that multiple people work on (and is a _target_, not a _source_ of PRs) merges should be preferred imo.

felixfbecker commented on src/vs/workbench/api/node/extHostLanguageFeatures.ts:
```diff
@@ -214,6 +214,33 @@ class ReferenceAdapter {
+		const convertReferences = (references: vscode.ReferenceInformation[]) => {
+			if (Array.isArray(references)) {
+				return references.map(reference => {
+					return <modes.IReferenceInformation>{
```
don't think the cast is needed
you can write this shorter without a `return`

nicksnyder replied:
> you can write this shorter without a return

Not sure what you mean? If I just delete the return I get a compile error.

> don't think the cast is needed

It isn't needed, but I think it is nice for readability since it is just a plain object.

felixfbecker replied:
In TypeScript /JavaScript arrow functions work like lambda expressions in functional languages, meaning without curly braces it returns an expression:
```ts
[1, 2, 3].map(x => x * 2) // [2, 4, 6]

fetch('whatever').then(res => res.json())

// note the paranthesis here so the object literal is not treated as a function body
references.map(reference => ({
  foo: reference.bar
}))
```

felixfbecker commented on src/vs/workbench/api/node/mainThreadLanguageFeatures.ts:
```diff
@@ -134,6 +134,28 @@ export class MainThreadLanguageFeatures extends MainThreadLanguageFeaturesShape
+				return wireCancellationToken(token, refs);
+			}
+		}, workspace);
+		return undefined;
```
The method is typed as returning a `Promise`

nicksnyder replied:
This is what all the other methods do, just trying to be consistent.

felixfbecker replied:
I am pretty sure the other code was written before TypeScript supported `undefined` types and strict null checks and they just didn't refactor everything. Imo new code should be typed correctly.

felixfbecker commented on src/vs/workbench/api/node/mainThreadLanguageFeatures.ts:
```diff
@@ -134,6 +134,28 @@ export class MainThreadLanguageFeatures extends MainThreadLanguageFeaturesShape
+		if (progressHandler) {
+			progressHandler(locations);
+		}
+		return undefined;
```
same

felixfbecker reviewed (changes requested):
I'm a bit confused why we integrate this in VS Code, aren't we only showing this in our custom sidebar in our own React code?

felixfbecker commented on src/vs/editor/contrib/referenceSearch/common/referenceSearch.ts:
```diff
@@ -40,7 +41,17 @@ export function provideReferences(model: IReadOnlyModel, position: Position, pro
+export function provideWorkspaceReferences(modeId: string, workspace: URI, query: ISymbolDescriptor, hints: { [hint: string]: any }, progress: (references: IReferenceInformation[]) => void): TPromise<IReferenceInformation[]> {
+
+	const model = {
+		isTooLargeForHavingARichMode: () => {
```
you can write this as
```ts
isTooLargeForHavingARichMode() {
  return false;
}
```

nicksnyder commented:
@rothfels ready for review, ptal

> I'm a bit confused why we integrate this in VS Code, aren't we only showing this in our custom sidebar in our own React code?

Supporting external references is something that is potentially upstreamable. Even if it weren't, it is nice to have external references going through the same pipeline as local refs.

nicksnyder committed:
Expose ReferenceContext so Sourcegraph can exclude dec

https://github.com/sourcegraph/sourcegraph/issues/4340

nicksnyder committed:
Support workspace/xreferences

https://github.com/sourcegraph/language-server-protocol/blob/master/extension-workspace-references.md
